arroyo-state/src/tables.rs

```diff
@@ -1,10 +1,85 @@
-//use crate::parquet::ParquetBackend;
-use crate::{BackingStore, StateBackend};
+use crate::{hash_key, BackingStore, DataOperation, StateBackend};
 use arroyo_rpc::grpc::{CheckpointMetadata, TableDescriptor, TableType};
 use arroyo_types::{from_micros, Data, Key, TaskInfo};
 use std::collections::{BTreeMap, HashMap, HashSet};
 use std::time::{Duration, SystemTime};
 
+pub enum Compactor {
+    TimeKeyMap,
+    TimeKeyMultiMap,
+}
+
+pub struct DataTuple<K, V> {
+    pub timestamp: SystemTime,
+    pub key: K,
+    pub value: Option<V>,
+    pub operation: DataOperation,
+}
+
+/// BlindDataTuple's key and value are not decoded
+#[derive(Debug)]
+pub struct BlindDataTuple {
+    pub key_hash: u64,
+    pub timestamp: SystemTime,
+    pub key: Vec<u8>,
+    pub value: Vec<u8>,
+    pub operation: DataOperation,
+}
+
+impl Compactor {
+    pub(crate) fn for_table_type(table_type: TableType) -> Self {
+        match table_type {
+            TableType::Global | TableType::TimeKeyMap => Compactor::TimeKeyMap,
+            TableType::KeyTimeMultiMap => Compactor::TimeKeyMultiMap,
+        }
+    }
+
+    pub(crate) fn compact_tuples(&self, tuples: Vec<BlindDataTuple>) -> Vec<BlindDataTuple> {
+        let mut result = vec![];
+        match self {
+            Compactor::TimeKeyMap => {
+                let mut data: BTreeMap<SystemTime, HashMap<Vec<u8>, Vec<u8>>> = BTreeMap::new();
+
+                for tuple in tuples {
+                    match tuple.operation {
+                        DataOperation::Insert => {
+                            data.entry(tuple.timestamp)
+                                .or_default()
+                                .insert(tuple.key, tuple.value);
+                        }
+                        DataOperation::DeleteKey => {
+                            // for each timestamp, remove the key
+                            for (_, map) in data.iter_mut() {
+                                map.remove(&tuple.key);
+                            }
+                        }
```

Comment:
@jacksonrnewhouse This is the main compaction logic for time key maps, please take a look.